standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/PersistenceManagerProvider.java

```diff
@@ -319,9 +327,17 @@ private static PersistenceManagerFactory initPMF(Configuration conf, boolean for
         LOG.warn("Could not create PersistenceManagerFactory using "
             + "connection pool properties, will fall back", e);
         pmf = JDOHelper.getPersistenceManagerFactory(prop);
+      } finally {
+        if (pmf == null && ds instanceof AutoCloseable) {
+          try (AutoCloseable close1 = (AutoCloseable) ds;
+               AutoCloseable close2 = (AutoCloseable) ds2 ) {
+          } catch (Exception e) {
+            LOG.warn("Failed to close the DataSource", e);
+          }
+        }
       }
     }
-    DataStoreCache dsc = pmf.getDataStoreCache();
+    DataStoreCache dsc = pmf != null ? pmf.getDataStoreCache() : null;
```

Comment:
warn would be misleading with that change when pmf is null
````
LOG.warn("PersistenceManagerFactory returned null DataStoreCache object
````

Reply:
yes, when reaching here the `pmf` shouldn't be null, the `finally` block will be executed whether there is a JDOFatalUserException or not, if in finally the `pmf` is null, there should be the exception thrown, adding the `return null` makes the impression this method can returning the nullable `pmf`, it makes little sense from my perspective.

Reply:
agree, but I don't see that you throw an exception in finally when `pmf` is null

Reply:
I write a simple code, 
```java
 public static void main(String[] args) throws Exception {
    String a = null;
    try {
     a();
    } catch (SQLException e) {
      throw new JDOFatalUserException();
    } finally {
      if (a == null) {
        System.out.println("print a");
      }
    }
    System.out.println("print b");
  }
  private static void a() throws Exception {
    throw new JDOFatalUserException();
  }
```
the output:
```
print a
Exception in thread "main" javax.jdo.JDOFatalUserException
	at org.example.FinallyException.a(FinallyException.java:23)
	at org.example.FinallyException.main(FinallyException.java:11)
```
as see, the finally is executed before the exception is thrown.

Reply:
@dengzhhu653  that's because you catch SQLException, but throw JDOException
````
  public  static void main(String ... args) {
    String a = null;
    try {
      a();
    } catch (Exception e) {
      System.out.println("catch block");
      throw new JDOFatalUserException("outer");
    } finally {
      System.out.println("finally");
      if (a == null) {
        System.out.println("print a");
      }
    }
    System.out.println("print b");
  }
  private static void a() throws Exception {
    throw new JDOFatalUserException("inner");
  } 
````
catch block
finally
print a
Exception in thread "main" javax.jdo.JDOFatalUserException: outer
	at org.apache.Main.main(Main.java:17)

Reply:
i think i get you point. `pmf` would be only null if  JDOFatalUserException is thrown, but we should still close the datasources